### ISSUE TITLE
elf::object::run_fini_funcs abort

### ISSUE THREAD

**Issue by justinc1** [cloudius-systems/osv#723]
Code and OS/gcc as in #722

Test below. With deafult "-c 1", no abort (10 runs). After -c8, abort in a few (n<10) trials.
```
sudo ./scripts/run.py -d -V -n -v -e '--ip=eth0,192.168.122.202,255.255.255.0 --defaultgw=192.168.122.1 --nameserver=192.168.122.1 /tools/mount-nfs.so nfs://192.168.122.1/ggg/?uid=0 /fff; tst-execve.so 30 /tst-file-read.so 30' --novnc -i build/debug/usr.img -c8

PP 29 tst-file-read.c ii=29/30
PP 28 tst-file-read.c ii=29/30
PP 25 tst-file-read.c ii=29/30
PP 20 tst-file-read.c ii=29/30
PP 3 tst-file-read.c main EXIT ret=0
PP 7 tst-file-read.c ii=29/30
PP 24 tst-file-read.c main EXIT ret=0
PP 1 fn=/fff/openfoam/etc/controlDict
0paPg e2 2f atusltt- foiultes-irdeea da.pcp lmiaciant ionE,X IaTd drre:t =00x
000000000000000
PP 26 tst-file-read.c main EXIT ret=0
PP 23 tst-file-read.c ii=29/30
[registers]
 s 1RI4 P:f 0nx0=00/00f00f00f07/2bo95pa e<sntdf::o_Rab_mtr/eee_itnscer/t_caned_lrlebMaloandcee(bloosl,
ed::_Rb_tree_node_base*, std::_Rb_tree_node_bPaPs e4* ,f ns=t/df:f:f_/Rob_tpreee_nondfeo_abma/etsec&/)+2c02>
llModels
RFPP 2 fn=/fff/Lopenfoam/etc/cellModels
: 0x0000P0P0 01030 0f0n1=0/2f8ff/6  oCpSe:n f o0axm0/0e0t0c0/0c0e000l000lModels
008  SSPP 18 fn=/fff/openfoam/etc/cellModels
e  0x0P0P0 01050 0f0n0=0/0f0f0f0/1o0p
 foam/etc/cellMoRdeAlXs:
0xffffa000051cb400  RBX: 0PPxffff a010020e 73f3cne0=  /RCfX:f 0fx0/00o00p0e0n0f0oam1/80e74t88c  /RDcX:e 0lx0l00M00o00d00e00ls0
0000
RPSI: 0xffffa0000217fd00  RDI: 0x00000000000000P0 01 0  RfBnP=:/ f0fxf0/0o0p0e2n0f00o03a4fmf8/50e  tR8c: / 0cxfefflfas00M00o51dcbe54l0

09:  0PxPf f1f7f af0n0=0/0f2fbf5/do7pbe0n f oRa1m0/:e t0cx/0c0e0l0l0M0o0d0e0l0s0
d000  R11: 0x000000000000000P0P   5 Rfn1=/2ff:f/ op0enxfo0am0/e0t0c2/0c0e00l34lffM8fof
els
013: 0xffffa00002b5d7b0  R1P4P:  106x 0f0n0=0/0f0f0f/0o0p0e0n0f1o0a0m0/ e tRc1/5c: e0xl00l0M0o0d0e0l0s0
000000  RSP: 0x00002000034ff808
 AbPor te8d
n
o[b/acfktfrafce/]
penfoam/etc/cellModels
PP 19 fn=/fff/openfoam/etc/controlDict
PP 6 fn=/fff/openfoam/etc/controlDict
ox000000000022d8f0 <abPoPr t0( cfhna=r/ fcfofn/sotp*e,n f.o.a.m)/+e2t4c9/>c
ntrolDict
PP 11 fn=/fff/openfoam/etc/controlDict
PP 21 fn=/fff/openfoam/etc/controlDict
PP 9 fn=/fff/openfoam/etc/controlDict
0x00000000003c7fdf <???+3964895>
0x00000000003c817e <mmu::vm_fault(unsigned long, exception_frame*)+350>
0x000000000048d37d <page_fault+315>
0x000000000048c206 <???+4768262>
0x000000000022efc8 <std::_Rb_tree_iterator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > > std::_Rb_tree<void*, std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > >, std::_Select1st<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > >, std::less<void*>, std::allocator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > > >::_M_emplace_hint_unique<std::piecewise_construct_t const&, std::tuple<void* const&>, std::tuple<> >(std::_Rb_tree_const_iterator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > >, std::piecewise_construct_t const&, std::tuple<void* const&>&&, std::tuple<>&&)+172>
0x000000000022ebd1 <std::map<void*, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > >, std::less<void*>, std::allocator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > > >::operator[](void* const&)+245>
0x000000000022da03 <__cxa_finalize+69>
0x0000103400000bc2 <???+3010>
0x00000000004039cd <elf::object::run_fini_funcs()+119>
0x000000000040531a <elf::program::remove_object(elf::object*)+140>
0x00000000004044c9 <???+4211913>
0x0000000000408829 <???+4229161>
0x000000000020e38f <std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()+65>
0x000000000020be8e <std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count()+38>
0x0000000000408ccd <std::__shared_ptr<elf::object, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr()+27>
0x0000000000408d0f <std::shared_ptr<elf::object>::~shared_ptr()+23>
0x0000000000643c9c <osv::application::~application()+156>
0x000000000064becd <void __gnu_cxx::new_allocator<osv::application>::destroy<osv::application>(osv::application*)+27>
0x000000000064beae <std::enable_if<std::__and_<std::allocator_traits<std::allocator<osv::application> >::__destroy_helper<osv::application>::type>::value, void>::type std::allocator_traits<std::allocator<osv::application> >::_S_destroy<osv::application>(std::allocator<osv::application>&, osv::application*)+34>
0x000000000064be7a <void std::allocator_traits<std::allocator<osv::application> >::destroy<osv::application>(std::allocator<osv::application>&, osv::application*)+34>
0x000000000064bd26 <std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()+54>
0x000000000020e38f <std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()+65>
0x000000000020be8e <std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count()+38>
0x000000000020ad6b <std::__shared_ptr<osv::application, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr()+27>
0x000000000020ad87 <std::shared_ptr<osv::application>::~shared_ptr()+23>
0x0000000000628da5 <???+6458789>
0x000000000062e622 <int std::_Bind_simple<int (*(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, long*, int))(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::__cxx11::basic_stri0x000000000062e3cc <std::_Bind_simple<int (*(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, long*, int))(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::__cxx11::basic_string<c0x000000000062e36b <std::thread::_Impl<std::_Bind_simple<int (*(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, long*, int))(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::__cx0x0000000000776b1f <execute_native_thread_routine+31>
0x00000000006a2003 <???+6955011>
0x000000000044fbef <std::function<void ()>::operator()() const+49>
0x00000000005c1f47 <sched::thread::main()+27>
0x00000000005be765 <thread_main_c+38>
0x000000000048d185 <???+4772229>
```

gdb trace:
```
(gdb) where
#0  processor::cli_hlt () at arch/x64/processor.hh:248
#1  0x0000000000209c54 in arch::halt_no_interrupts () at arch/x64/arch.hh:48
#2  0x000000000049d63b in osv::halt () at arch/x64/power.cc:24
#3  0x000000000022d917 in abort (fmt=0xa3705d "Aborted\n") at runtime.cc:132
#4  0x000000000022d7f7 in abort () at runtime.cc:96
#5  0x00000000003c7fe0 in mmu::vm_sigsegv (addr=0, ef=0xffff80000c008078) at core/mmu.cc:1315
#6  0x00000000003c817f in mmu::vm_fault (addr=0, ef=0xffff80000c008078) at core/mmu.cc:1337
#7  0x000000000048d37e in page_fault (ef=0xffff80000c008078) at arch/x64/mmu.cc:38
#8  <signal handler called>
#9  0x000000000072b95a in std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) ()
#10 0x000000000022fa24 in std::_Rb_tree<void*, std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > >, std::_Select1st<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > >, std::less<void*>, std::allocator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > > >::_M_insert_node (this=0x1807480 <destructors>, __x=0x0, __p=0xffffa0000e733cc0, __z=0xffffa0000217fd00)
    at /usr/include/c++/5.3.1/bits/stl_tree.h:2076
#11 0x000000000022efc9 in std::_Rb_tree<void*, std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > >, std::_Select1st<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > >, std::less<void*>, std::allocator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > > >::_M_emplace_hint_unique<std::piecewise_construct_t const&, std::tuple<void* const&>, std::tuple<> >(std::_Rb_tree_const_iterator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > >, std::piecewise_construct_t const&, std::tuple<void* const&>&&, std::tuple<>&&) (
    this=0x1807480 <destructors>, __pos=...) at /usr/include/c++/5.3.1/bits/stl_tree.h:2177
#12 0x000000000022ebd2 in std::map<void*, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > >, std::less<void*>, std::allocator<std::pair<void* const, std::vector<std::pair<void (*)(void*), void*>, std::allocator<std::pair<void (*)(void*), void*> > > > > >::operator[] (this=0x1807480 <destructors>, 
    __k=@0x2000034ff958: 0x103400201e00) at /usr/include/c++/5.3.1/bits/stl_map.h:483
#13 0x000000000022da04 in __cxxabiv1::__cxa_finalize (dso=0x103400201e00) at runtime.cc:175
#14 0x0000103400000bc3 in __do_global_dtors_aux ()
#15 0x00002000034ffa00 in ?? ()
#16 0x00000000004039ce in elf::object::run_fini_funcs (this=0x2000034ff9c0) at core/elf.cc:956
---Type <return> to continue, or q <return> to quit---
Backtrace stopped: frame did not save the PC
```

**Comment by justinc1** [comment]
Seems same as #713

**Comment by nyh** [comment]
Indeed appears a duplicate of #713? Because #713 did not, if I understood correctly, involve NFS, I think this bug is related to ELF namespaces, or something related - but not to NFS.

**Comment by nyh** [comment]
I just thought of something, when seeing frame 12 above: our code __cxa_atexit() does not protect the "destructors" map with a mutex! So if we have code where we load many objects in parallel, we can run into two threads trying to write into this map at the same and trashing it...

If I understand correctly, your use case does try to load many objects in parallel, because you don't wait for one to load before starting another thread to load another one, and so on?

I'll try to provide a patch tomorrow and ask you to test it.

**Comment by justinc1** [comment]
You are right, NFS is not common point between #713, #723. But osv_execve and ELF namespaces are used in both cases.

**Comment by justinc1** [comment]
Yes, in above case, I start from one thread 30 other threads. And it is in parallel - I even intentionally create additional thread, so that osv_execve doesn't wait on real payload to finish.
